src/epik_eval/dataloader.py

```diff
@@ -0,0 +1,205 @@
+                        train_doc_stories.append(' '.join(doc[1:]))
+                    else:
+                        train_doc_stories.append('\n'.join(doc[1:]))
+
+        self.prompts = questions + train_doc_banners
+        self.targets = answers + train_doc_stories
+
+        # Prepare batches once since always the same
+        self.batches = []
+        for i in range(0, len(self.prompts), batch_size):
+            batch = tokenizer(self.prompts[i: i + batch_size], padding=True, return_tensors='pt')
+            prompt_ids = range(i, min(i + batch_size, len(self.prompts)))
+            self.batches.append((prompt_ids, batch.input_ids, batch.attention_mask))
+
+        train_df = pd.DataFrame({
+            'set': ['train'] * len(train_doc_ids),
+            'story id': train_doc_ids,
+            'task': train_doc_tasks,
+            'question': train_doc_banners,
+            'answer': train_doc_stories
+        })
+        self.eval_df = pd.concat([qa_df, train_df], ignore_index=True)
+
+        self.current_batch = 0
+        self.finished = False
+
+    def __iter__(self):
+        if self.finished:
+            warnings.warn('EvaluationDataset.__iter__() was called, but all questions have already been iterated.')
+            return
+
+        prompt_ids, prompts, attention_mask = self.batches[self.current_batch]
+        for i in range(prompts.shape[0]):
+            yield prompt_ids[i], prompts[i], attention_mask[i]
+
+        self.current_batch += 1
+        self.finished = self.current_batch >= len(self.batches)
+
+    def __len__(self):
+        return len(self.questions)
+
+    def reset(self):
+        self.current_batch = 0
+        self.finished = False
+
+
+class TrainDataLoader(torch.utils.data.DataLoader):
+    """Wrapper for torch.utils.data.DataLoader. Is meant to be used with the TrainDataset, required by the Accelerate module, but then gets replaced with Accelerate's own dataloader."""
+
+    def __init__(self, tokenizer, model_type, segmented_docs, batch_size):
+        dataset = TrainDataset(tokenizer, model_type, segmented_docs)
+        super().__init__(dataset, batch_size=batch_size, shuffle=False)
+
+    def __iter__(self):
+        for batch in self.dataset:
+            yield batch
+
+
+class EvaluationDataLoader(torch.utils.data.DataLoader):
+    """Wrapper for torch.utils.data.DataLoader. Is meant to be used with the QuestionDataset, required by the Accelerate module, but then gets replaced with Accelerate's own dataloader."""
+
+    def __init__(self, tokenizer, segmented_docs, batch_size, model_type):
+        super().__init__(
+            EvaluationDataset(
+                tokenizer,
+                batch_size,
+                model_type,
+                segmented_docs),
+            batch_size=batch_size,
+            shuffle=False)
+
+    def __iter__(self):
+        for batch in self.dataset:
```

Comment:
`return iter(self.dataset)` instead. Avoids keeping a generator scope and object.

Reply:
This is placeholder code, as the dataloader gets replaced by Accelerate's.